### ISSUE TITLE
`redundant_self` false positive in 0.63.0-rc.2 (escaping autoclosure)

### ISSUE THREAD

**Issue by shagedorn** [realm/SwiftLint#6389]
### New Issue Checklist

- [x] I've Updated SwiftLint to the latest version.
- [x] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

Swift requires explicit self usage in escaping closures. The new `redundant_self` rule doesn't catch this for autoclosures.

Assuming this helper function:

```
func checkCondition(
  _ condition: @autoclosure @escaping () -> Bool,
) { … }
```

A valid call needs to look like this:

```
checkCondition(self.myState)
```

The new linter rule suggests;

```
checkCondition(myState)
```

…but that throws a Swift error:

> Reference to property 'myState' in closure requires explicit use of 'self' to make capture semantics explicit

### Environment

* SwiftLint version: 0.63.0-rc.2
* Xcode version: 26.2
* Installation method used: release ZIP file download
* Configuration file: irrelevant, except that `redundant_self` is enabled

**Comment by SimplyDanny** [comment]
I fear this is a case we cannot fix. SwiftLint, operating on the AST without type information only, cannot see the declaration of `checkCondition` and thus doesn't know about the `@autoclosure` attribute.

**Comment by shagedorn** [comment]
That's sad but understandable, thank you for explaining 🙏🏻

What I noticed when playing with the new (and greatly appreciated 🙌🏻) `redundant_self` rule is that it replaces `redundant_self_in_closure`, but doesn't actually allow to use (only) the old behaviour, unless I'm missing something.

I could imagine that many codebases were fine with `redundant_self_in_closure`, but cannot adopt `redundant_self` (_at least not for every directory/module_) due to either preferences or false positives like this one. Would you consider adding an option such as `only_in_closures` to essentially pin down the previous behaviour?

Our concrete use case is this: `redundant_self` does very well in our codebase, except for unit test modules, because some of our test helpers heavily use autoclosures.

We already have tooling in place that puts a refined linter config into all test module directories, so I'd like to enable `redundant_self` with the `keep_in_initializers` option in the root config, and overwrite it in test configs by adding the `only_in_closures` option to work around the autoclosure limitation without completely giving up on the rule.

**Comment by SimplyDanny** [comment]
I thought about an option when I opened up the rule, but then decided against it because of unclear reasons why people would only like to avoid `self.` in closures. So I went with the pre-release just widening the rule. Apparently, there are reasons and adding an option should not be a big deal.
